src/dc_tools.rs

```diff
@@ -616,43 +617,79 @@ pub unsafe fn dc_create_incoming_rfc724_mid(
     )
 }
 
-pub unsafe fn dc_create_outgoing_rfc724_mid(
-    grpid: *const libc::c_char,
-    from_addr: *const libc::c_char,
-) -> *mut libc::c_char {
+pub fn dc_create_outgoing_rfc724_mid(
+    context: &Context,
+    group_id: Option<&str>,
+    from_addr: &str,
+) -> String {
     /* Function generates a Message-ID that can be used for a new outgoing message.
     - this function is called for all outgoing messages.
     - the message ID should be globally unique
     - do not add a counter or any private data as as this may give unneeded information to the receiver	*/
-    let mut rand1: *mut libc::c_char = ptr::null_mut();
-    let rand2: *mut libc::c_char = dc_create_id().strdup();
-    let ret: *mut libc::c_char;
-    let mut at_hostname: *const libc::c_char = strchr(from_addr, '@' as i32);
-    if at_hostname.is_null() {
-        at_hostname = b"@nohost\x00" as *const u8 as *const libc::c_char
-    }
-    if !grpid.is_null() {
-        ret = dc_mprintf(
-            b"Gr.%s.%s%s\x00" as *const u8 as *const libc::c_char,
-            grpid,
-            rand2,
-            at_hostname,
-        )
+
+    let at_hostname = match from_addr.find('@') {
+        None => "@nohost",
+        Some(pos) => &from_addr[pos..]
+    };
+
+    match group_id {
+        None => {
+            let mid_prefix = match context.get_config(Config::Rfc724MsgIdPrefix) {
+                Some(ref s) if s == "1" => "chat$",
+                _ => "Mr."
```

Comment:
i'd usually write None instead of "_" here.

Reply:
But `None` would be wrong as there is an `if s == "1"` on the Some match.